File: forms/birth-control-assessment/back-end-with-loco/src/engine/mec_grader.rs
//! Mec grader module.

use super::flagged_issues::detect_additional_flags;
use super::mec_rules::all_rules;
use super::types::{AssessmentData, FiredRule, GradingResult, MethodMEC};

/// Pure function: grades a birth-control assessment against the UK MEC rules.
///
/// Returns a [`GradingResult`] containing:
/// - per-method MEC categories (worst category wins),
/// - overall risk level derived from the worst MEC category,
/// - the rules that fired,
/// - additional flagged issues independent of MEC classification.
pub fn calculate_mec_grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let fired_rules = run_rules(data);
    let method_mec = derive_method_mec(&fired_rules);
    let overall_risk = derive_overall_risk(&method_mec);
    let additional_flags = detect_additional_flags(data);

    GradingResult {
        method_mec,
        overall_risk,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

/// Evaluate every rule against the data, returning the rules that fired.
pub fn run_rules(data: &AssessmentData) -> Vec<FiredRule> {
    let mut out = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            out.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                mec_category: rule.mec_category,
                affected_methods: rule
                    .affected_methods
                    .iter()
                    .map(|s| s.to_string())
                    .collect(),
            });
        }
    }
    out
}

/// Derive per-method MEC categories from fired rules (worst wins).
fn derive_method_mec(fired_rules: &[FiredRule]) -> MethodMEC {
    let mut result = MethodMEC::default();
    let methods = ["coc", "pop", "implant", "injection", "iud", "ius"];

    for rule in fired_rules {
        for method in &methods {
            if rule.affected_methods.iter().any(|m| m == method) {
                let current = match *method {
                    "coc" => &mut result.coc,
                    "pop" => &mut result.pop,
                    "implant" => &mut result.implant,
                    "injection" => &mut result.injection,
                    "iud" => &mut result.iud,
                    "ius" => &mut result.ius,
                    _ => continue,
                };
                if rule.mec_category > *current {
                    *current = rule.mec_category;
                }
            }
        }
    }
    result
}

/// Derive overall risk level from worst MEC category across all methods.
fn derive_overall_risk(method_mec: &MethodMEC) -> String {
    let worst = [
        method_mec.coc,
        method_mec.pop,
        method_mec.implant,
        method_mec.injection,
        method_mec.iud,
        method_mec.ius,
    ]
    .into_iter()
    .max()
    .unwrap_or(1);

    match worst {
        n if n >= 4 => "critical".to_string(),
        3 => "high".to_string(),
        2 => "moderate".to_string(),
        _ => "low".to_string(),
    }
}

File: forms/birth-control-assessment/front-end-with-html/js/mec-grader.js
import { detectAdditionalFlags } from './flagged-issues.js';
import { mecRules } from './mec-rules.js';

// UK MEC grader. Pure functions: take an `AssessmentData` object, evaluate
// every MEC rule, derive per-method MEC categories (worst category wins) and
// an overall risk level. Mirrors `src/lib/engine/mec-grader.ts` from the
// SvelteKit reference.

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').FiredRule} FiredRule
 * @typedef {import('./types.js').MethodMEC} MethodMEC
 * @typedef {import('./types.js').MECCategory} MECCategory
 * @typedef {import('./types.js').RiskLevel} RiskLevel
 * @typedef {import('./types.js').GradingResult} GradingResult
 */

/**
 * Evaluate every MEC rule and return the grading result.
 * @param {AssessmentData} data
 * @returns {GradingResult}
 */
function calculateMECGrade(data) {
  /** @type {FiredRule[]} */
  const firedRules = [];

  for (const rule of mecRules) {
    try {
      if (rule.evaluate(data)) {
        firedRules.push({
          id: rule.id,
          category: rule.category,
          description: rule.description,
          mecCategory: rule.mecCategory,
          affectedMethods: rule.affectedMethods.slice()
        });
      }
    } catch (e) {
      console.warn(`MEC rule ${rule.id} evaluation failed:`, e);
    }
  }

  const methodMEC = deriveMethodMEC(firedRules);
  const overallRisk = deriveOverallRisk(methodMEC);
  const additionalFlags = detectAdditionalFlags(data);

  return {
    methodMEC,
    overallRisk,
    firedRules,
    additionalFlags,
    timestamp: new Date().toISOString()
  };
}

/**
 * Per-method MEC categories: worst category wins. Methods not affected by any
 * fired rule default to MEC 1.
 * @param {FiredRule[]} firedRules
 * @returns {MethodMEC}
 */
function deriveMethodMEC(firedRules) {
  /** @type {(keyof MethodMEC)[]} */
  const methods = ['coc', 'pop', 'implant', 'injection', 'iud', 'ius'];
  /** @type {MethodMEC} */
  const result = { coc: 1, pop: 1, implant: 1, injection: 1, iud: 1, ius: 1 };

  for (const rule of firedRules) {
    for (const method of methods) {
      if (rule.affectedMethods.includes(method)) {
        if (rule.mecCategory > result[method]) {
          result[method] = rule.mecCategory;
        }
      }
    }
  }

  return result;
}

/**
 * Map the worst MEC category across methods to a risk level.
 * @param {MethodMEC} methodMEC
 * @returns {RiskLevel}
 */
function deriveOverallRisk(methodMEC) {
  const worstMEC = Math.max(
    methodMEC.coc,
    methodMEC.pop,
    methodMEC.implant,
    methodMEC.injection,
    methodMEC.iud,
    methodMEC.ius
  );
  if (worstMEC >= 4) return 'critical';
  if (worstMEC >= 3) return 'high';
  if (worstMEC >= 2) return 'moderate';
  return 'low';
}

export { calculateMECGrade, deriveMethodMEC, deriveOverallRisk };
